docs(blog): New Blogpost (#348)
AndreaPontrandolfo/sheriff: apps/docs-website/blog/2025-01-5-inversion-of-control.mdx

:::info

I know, Inversion of Control is an expression that can have many different meanings in computer programming, depending on the context, but in this article I'm using it in loose terms to describe the process of giving the user more control over the library's behavior.

:::

## Why

One of the coolest things about Sheriff is the way it encapsulates the complexity of ESLint and its ecosystem, most of the times hiding the ugly details behind the curtains. The problem is that sometimes the level of abstraction is too intrusive and can obstacolate the user's ability to customize the configuration to their needs. It's a delicate balance to strike.

In this article we will explore why the Inversion of Control principle is often the key to unlock the best solution.

## The `noRestrictedSyntaxOverride` option

### The problem

ESLint offers a rule called [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) that allows you to disallow specific Javascript syntax features.

Sheriff comes with a preconfigured `no-restricted-syntax` rule that disallows some of the most common syntax features that are considered harmful or confusing.

> The problem is that the rule accepts an array of features, meaning that if the user doesn't like just one of these features, they had only 2 options:

Reviewer: ```suggestion
The problem here is that the rule accepts an array of features, meaning that if the user dislike any one of these lints, they had only 2 options:
```

> - disable the rule entirely
> 
>   ```ts
>   noRestrictedSyntax: 0
>   ```
> - override the rule with a new one
> 
>   ```ts
>   noRestrictedSyntax: [
>     2,
>     {
>       selector: "...",
>       message: "...",
>     },
>     {
>       selector: "...",
>       message: "...",
>     },
>   ]
>   ```

Reviewer: ````suggestion
1. disable the rule entirely

   ```ts
   noRestrictedSyntax: 0
   ```
2.  override the rule with a new one

   ```ts
   noRestrictedSyntax: [
     2,
     {
       selector: "...",
       message: "...",
     },
     {
       selector: "...",
       message: "...",
     },
   ]
   ```
````

> So in Sheriff we came up with a solution that allows the user to disable just the specific feature they don't like or append new ones.

Reviewer: ```suggestion
In Sheriff, we came up with a solution that allows users to both disable specific restrictions they disliked and append new ones.
```

The API looked like this:

```ts
noRestrictedSyntaxOverride: {
    adjuncts: [
        {
            selector: "LabeledStatement",
            message: "...",
        },
        {
            selector: "ForInStatement",
            message: "...",
        },
    ],
    allows: [
        "LabeledStatement",
        "ForInStatement",
    ],
},
```

> In the `adjuncts` array, the user can add new syntax features to disallow on-top of the default ones, while in the `allows` array they can disable some of the default ones.

Reviewer: ```suggestion
In the `adjuncts` array, users can add syntax features to disallow on top of the defaults, while they can use the `allows` array to disable specific defaults.
```

> As you can see, the offered API looked pretty complex, arbitrary, tangled and unintuitive. Both `adjuncts` and `allows` suggest at "adding" something, so it could be confusing to understand "what did what" for newcomers of the library. In short: it was not user-friendly at all.

Reviewer: ```suggestion
As you can see, the offered API looked rather complex and unintuitive. Both `adjuncts` and `allows` suggest "adding" something, so newcomers found understanding "what did what" difficult. In short: it was not user-friendly at all.
```

### The solution

> Inversion of control to the rescue.

Reviewer: ```suggestion
Inversion of control to the rescue!
```

> Instead of wrapping the `no-restricted-syntax` option into the Sheriff options, we let the users define and configure the `no-restricted-syntax` rule directly in their own config and offer them **composables** around it. This way the user is back in the driver seat and has full control again.

Reviewer: ```suggestion
Instead of wrapping `no-restricted-syntax` options into Sheriff’s options, we let users define and configure the `no-restricted-syntax` rule directly in their own config and then offer **composables** around it. This way, our users are back in the driver’s seat and have full control once more.
```

> Sheriff now expose a variable called `baseNoRestrictedSyntaxRules` that contains the contents of the Sheriff-configured `no-restricted-syntax` rule. So adding new restricted features on top of the default ones is as simple as:

Reviewer: ```suggestion
Sheriff now expose a variable called `baseNoRestrictedSyntaxRules` that contains the contents of the Sheriff-configured `no-restricted-syntax` rule. With this simple API, adding new restrictions on top of the defaults is as simple as:

```ts
no-restricted-syntax: [
    2,
    ...baseNoRestrictedSyntaxRules,
    // your custom rules here...
],
```

> While disabling some of the default ones is a little more involved.

Reviewer: ```suggestion
Admittingly, disabling the defaults is a little more involved.
```

> Every entry prints the index of itself at the end of the message property, so users should use the index to identify which entry they want to remove.

Reviewer: ```suggestion
Each entry prints its index at the end of the message, so users can use the indices to identify which entries they wish to remove.
```

> In the docs we suggest using the native Javascript [`toSpliced()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSpliced) method to remove an entry from the `baseNoRestrictedSyntaxRules` array:

Reviewer: ```suggestion
In the docs we suggest using the native JavaScript [`Array#toSpliced()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSpliced) method to remove an entry from the `baseNoRestrictedSyntaxRules` array:
```

```ts
no-restricted-syntax: [
    2,
    ...baseNoRestrictedSyntaxRules.toSpliced(2, 1)
    // your custom rules here...
],
```

[🠪 Learn more about this API in the docs](https://www.eslint-config-sheriff.dev/docs/configuration#customizing-the-no-restricted-syntax).

### Going forward

> Recently an ESlint contributor released a very interesting alternative to the `no-restricted-syntax` rule called [eslint-no-restricted](https://github.com/bradzacher/eslint-no-restricted).

Reviewer: ```suggestion
Recently, Brad Zacher released an interesting alternative to `no-restricted-syntax` called [eslint-no-restricted](https://github.com/bradzacher/eslint-no-restricted).
```

> Sheriff will most likely adopt this library in the future and abandon the current "homegrown" solution, as it seems to provide a superior DX.

Reviewer: ```suggestion
Sheriff will likely adopt this library in the future and abandon the current, homegrown solution, as Brad’s solution seems to provide superior DX.
```

## Typescript-eslint `project` API

> Another case where Sheriff was swallowing up complexity at the cost of giving up a level of control to the user was the `parserOptions.project` API.

Reviewer: ```suggestion
Another case where Sheriff swallowed up complexity at the cost of taking away control from users was with typescript-eslint’s `parserOptions.project` API.
```

### The problem

> Sheriff was passing the `tsconfig.json` to the ESLint config in a hard-coded way:

Reviewer: ```suggestion
Sheriff used to hardcode the path to the `tsconfig.json` it passed to typescript-eslint.
```

```ts
project: './tsconfig.json'
```

> Naturally users with multiple `tsconfig.json` on the same level were feeling limited, because it wasn't possible to specify different paths for the `tsconfig.json`.

Reviewer: ```suggestion
Naturally, users with multiple `tsconfig.json`s felt limited because it wasn't possible to specify different paths for `tsconfig.json`s.
```

### The solution

The solution was to let the user define the `project` path through the Sheriff options:

```ts
pathsOverrides: {
    tsconfigLocation: "./tsconfig.sample.json",
},
```

> and fallback to the [default one](https://typescript-eslint.io/blog/parser-options-project-true/) if not specified:

Reviewer: ```suggestion
and fall back to the [default](https://typescript-eslint.io/blog/parser-options-project-true/) if not specified:
```

```ts
parserOptions: {
    project: userChosenTSConfig || true,
},
```

### Going forward

> Since then the typescript-eslint team moved to a new option called [`projectService`](https://typescript-eslint.io/blog/announcing-typescript-eslint-v8-beta#project-service).

Reviewer: ```suggestion
Since then, the typescript-eslint team has moved to a new system called the [project service](https://typescript-eslint.io/blog/announcing-typescript-eslint-v8#project-service).
```

> This option is meant to definitely get rid of the need for the custom tsconfigs, like `tsconfig.eslint.json` etc, that some users had for some advanced use-cases. It also has other benefits. [Learn more](https://typescript-eslint.io/packages/parser#projectservice).

Reviewer: ```suggestion
This option is meant to get rid of the need for custom `tsconfig.json`s like `tsconfig.eslint.json`, etc., that some users had made for advanced use cases. The project service also has other benefits, and is often more performant. [Learn more](https://typescript-eslint.io/packages/parser#projectservice).
```

> Sheriff is currently exploring on adopting this new API ([Issue](https://github.com/AndreaPontrandolfo/sheriff/issues/203) | [PR](https://github.com/AndreaPontrandolfo/sheriff/pull/205)), but it seems like the feature is still experimental and [has some issues to iron out first](https://github.com/microsoft/vscode-eslint/issues/1911).

Reviewer: ```suggestion
Sheriff is currently exploring on adopting this new API ([issue](https://github.com/AndreaPontrandolfo/sheriff/issues/203) | [PR](https://github.com/AndreaPontrandolfo/sheriff/pull/205)).
```

The linked issue states that the project service is now more correct than the old version (in `typescript-eslint` `>=8.3.0`), although the issue remains open until issues (that also exist with the current version) are fixed. It's also no longer experimental in typescript-eslint v8, but has been stabilized.

## Conclusion

> By simplifying APIs and empowering advanced users with more granular controls, Sheriff strikes a balance between ease of use and developer freedom. As the library evolves, it remains committed to enhancing developer experience with modern, user-focused solutions.

Reviewer: ```suggestion
By simplifying APIs and empowering users with more granular controls, Sheriff seeks to strike a balance between ease of use and developer freedom. As the library evolves, we remain committed to enhancing developer experience with modern, user-focused solutions.
```
